utils/zerovec/src/varzerovec/components.rs

```diff
@@ -0,0 +1,243 @@
+            .get(8..8 * len + 8)
+            .ok_or(VarZeroVecError::FormatError)?;
+        let indices = PlainOldULE::<8>::parse_byte_slice(indices_bytes)
+            .map_err(|_| VarZeroVecError::FormatError)?;
+        let things = slice
+            .get(8 * len + 8..)
+            .ok_or(VarZeroVecError::FormatError)?;
+
+        let components = SliceComponents {
+            indices,
+            things,
+            entire_slice: slice,
+            marker: PhantomData,
+        };
+
+        for thing in components.iter_checked() {
+            let _ = thing?;
+        }
+
+        Ok(components)
+    }
+
+    #[inline]
+    pub fn len(self) -> usize {
+        self.indices.len()
+    }
+
+    #[inline]
+    #[cfg(feature = "serde")]
+    pub fn entire_slice(self) -> &'a [u8] {
+        self.entire_slice
+    }
+
+    /// Get the idx'th element out of this slice. Returns `None` if out of bounds.
+    #[inline]
+    pub fn get(self, idx: usize) -> Option<&'a T::VarULE> {
+        if idx >= self.len() {
+            return None;
+        }
+        Some(unsafe { self.get_unchecked(idx) })
+    }
+
+    /// Get the idx'th element out of this slice. Does not bounds check.
+    ///
+    /// Safety:
+    /// - `idx` must be in bounds (`idx < self.len()`)
+    #[inline]
+    unsafe fn get_unchecked(self, idx: usize) -> &'a T::VarULE {
+        let start = usizeify(*self.indices.get_unchecked(idx));
+        let end = if idx + 1 == self.len() {
+            self.things.len()
+        } else {
+            usizeify(*self.indices.get_unchecked(idx + 1))
+        };
+        debug_assert!(start <= end);
+        let things_slice = self.things.get_unchecked(start..end);
+        T::VarULE::from_byte_slice_unchecked(things_slice)
+    }
+
+    /// Create an iterator over the Ts contained in SliceComponents, checking internal invariants:
+    ///
+    /// - `indices[i]..indices[i+1]` must index into a valid section of
+    ///   `things`, such that it parses to a `T::VarULE`
+    /// - `indices[len - 1]..things.len()` must index into a valid section of
+    ///   `things`, such that it parses to a `T::VarULE`
+    /// - `indices` is monotonically increasing
+    ///
+    /// This method is NOT allowed to call any other methods on SliceComponents since all other methods
+    /// assume that the slice has been passed through iter_checked
+    #[inline]
+    fn iter_checked(self) -> impl Iterator<Item = Result<&'a T::VarULE, ParseErrorFor<T>>> {
+        let last = iter::from_fn(move || {
+            if self.len() > 0 {
+                let start = usizeify(self.indices[self.len() - 1]);
+                let end = self.things.len();
+                Some(
+                    self.things
+                        .get(start..end)
+                        .ok_or(VarZeroVecError::FormatError),
+                )
+            } else {
+                None
+            }
+        })
+        .take(1);
+        self.indices
+            .windows(2)
```

Comment:
Suggestion (optional): make the `.chain(last)` add in the indices tuple, rather than the slice.  Then you can eliminate the duplicate `self.things.get(...).ok_or(...)` above.

Reply:
We can't, `.windows()` is implemented on slices, not iterators